packages/shared/src/open-feature.ts

```diff
@@ -1,21 +1,208 @@
+    }
+
+    const clientEmitter = this.getEventEmitterForClient(clientName);
+    this.transferListeners(oldProvider, provider, clientName, clientEmitter);
+
+    if (typeof provider.initialize === 'function') {
+      provider
+        .initialize?.(this._context)
+        ?.then(() => {
+          clientEmitter.emit(ProviderEvents.Ready, { clientName });
+          this._events?.emit(ProviderEvents.Ready, { clientName });
+        })
+        ?.catch((error) => {
+          clientEmitter.emit(ProviderEvents.Error, { clientName, message: error.message });
+          this._events?.emit(ProviderEvents.Error, { clientName, message: error.message });
+        });
+    } else {
+      clientEmitter.emit(ProviderEvents.Ready, { clientName });
+      this._events?.emit(ProviderEvents.Ready, { clientName });
+    }
+
+    // Do not close the default provider if a named client used the default provider
+    if (!clientName || (clientName && oldProvider !== this._defaultProvider)) {
+      oldProvider?.onClose?.();
+    }
+    return this;
+  }
+
+  protected getProviderForClient(name?: string): P {
+    if (!name) {
+      return this._defaultProvider;
+    }
+
+    return this._clientProviders.get(name) ?? this._defaultProvider;
+  }
+
+  protected getEventEmitterForClient(name?: string): OpenFeatureEventEmitter {
+    const emitter = this._clientEvents.get(name);
+
+    if (emitter) {
+      return emitter;
+    }
+
+    const newEmitter = new OpenFeatureEventEmitter(() => this._logger);
+    this._clientEvents.set(name, newEmitter);
+    return newEmitter;
+  }
+
+  private transferListeners(
+    oldProvider: P,
+    newProvider: P,
+    clientName: string | undefined,
+    clientEmitter: OpenFeatureEventEmitter
+  ) {
+    oldProvider.events?.removeAllHandlers();
+
+    // iterate over the event types
+    Object.values<ProviderEvents>(ProviderEvents).forEach((eventType) =>
+      newProvider.events?.addHandler(eventType, async (details?: EventDetails) => {
+        // on each event type, fire the associated handlers
+        clientEmitter.emit(eventType, { ...details, clientName });
+        this._events.emit(eventType, { ...details, clientName });
+      })
+    );
+  }
+
+  async close(): Promise<void> {
+    try {
+      await this?._defaultProvider?.onClose?.();
+    } catch (err) {
+      this.handleShutdownError(this._defaultProvider, err);
+    }
+
+    const providers = Array.from(this._clientProviders);
+
+    await Promise.all(
+      providers.map(async ([, provider]) => {
+        try {
+          await provider.onClose?.();
+        } catch (err) {
+          this.handleShutdownError(this._defaultProvider, err);
+        }
+      })
+    );
+  }
+
+  private handleShutdownError(provider: P, err: unknown) {
+    this._logger.error(`Error during shutdown of provider ${provider.metadata.name}: ${err}`);
+
+    if (err instanceof Error) {
+      this._logger.error(err.stack);
+    }
+    this._logger.error((err as Error)?.stack);
```

Comment:
This seems odd. We are logging the stack no matter what, so no need to check I think?

Apologies if I wrote this and you moved it, not sure. :sweat_smile:

Reply:
That is true😅
And yes it is true I moved it from the other place.

Reply:
Changed!